File: server/Simulation/Components/SpawnerComponent.cs
using System;
using server.ECS;
using server.Enums;

namespace server.Simulation.Components;

[Component(ComponentType = ComponentType.Spawner, NetworkSync = true)]
public struct SpawnerComponent(int unitId, TimeSpan interval, int amountPerInterval, int maxPopulation, int minPopulation)
{
    public long ChangedTick = NttWorld.Tick;
    public readonly int UnitIdToSpawn = unitId;

    // Spawn Interval 
    public readonly TimeSpan Interval = interval;

    public float TimeSinceLastSpawn = 0;

    // Amount of UnitType to spawn each Interval
    public readonly int AmountPerInterval = amountPerInterval;

    // Limit of living UnitIdToSpawn after which Interval will be ignored
    public readonly int MaxPopulation = maxPopulation;

    // The idea is that when there's less than MinPopulation of living UnitIdToSpawn,
    // Interval will be ignored and units will be spawned each frame until 
    // MinPopulation of alive UnitIdToSpawn has been reached
    public readonly int MinPopulation = minPopulation;


}

File: pixiejsClient/src/app/ecs/components/SpawnerComponent.ts
import { Component, component, serverField } from "../core/Component";
import { NTT } from "../core/NTT";
import { ServerComponentType } from "../../enums/ComponentIds";

@component(ServerComponentType.Spawner)
export class SpawnerComponent extends Component {
  // changedTick is inherited from Component base class
  @serverField(1, "i32") unitIdToSpawn: number;
  @serverField(2, "i64") interval: bigint; // TimeSpan as ticks
  @serverField(3, "f32") timeSinceLastSpawn: number;
  @serverField(4, "i32") amountPerInterval: number;
  @serverField(5, "i32") maxPopulation: number;
  @serverField(6, "i32") minPopulation: number;

  constructor(
    ntt: NTT,
    unitIdToSpawn: number = 0,
    interval: bigint = 0n,
    timeSinceLastSpawn: number = 0,
    amountPerInterval: number = 1,
    maxPopulation: number = 10,
    minPopulation: number = 0,
  ) {
    super(ntt);
    this.unitIdToSpawn = unitIdToSpawn;
    this.interval = interval;
    this.timeSinceLastSpawn = timeSinceLastSpawn;
    this.amountPerInterval = amountPerInterval;
    this.maxPopulation = maxPopulation;
    this.minPopulation = minPopulation;
  }
}
